src/measure_theory/function/strongly_measurable.lean

```diff
@@ -0,0 +1,345 @@
+⟨hf.approx, hf.tendsto_approx⟩
+
+lemma exists_set_sigma_finite [topological_space β] [t2_space β]
+  (hf : fin_strongly_measurable f μ) :
+  ∃ t, measurable_set t ∧ (∀ x ∈ tᶜ, f x = 0) ∧ sigma_finite (μ.restrict t) :=
+begin
+  rcases hf with ⟨fs, hT_lt_top, h_approx⟩,
+  let T := λ n, support (fs n),
+  have hT_meas : ∀ n, measurable_set (T n), from λ n, simple_func.measurable_set_support (fs n),
+  let t := ⋃ n, T n,
+  refine ⟨t, measurable_set.Union hT_meas, _, _⟩,
+  { have h_fs_zero : ∀ n, ∀ x ∈ tᶜ, fs n x = 0,
+    { intros n x hxt,
+      rw [set.mem_compl_iff, set.mem_Union, not_exists] at hxt,
+      simpa using (hxt n), },
+    refine λ x hxt, tendsto_nhds_unique (h_approx x) _,
+    rw funext (λ n, h_fs_zero n x hxt),
+    exact tendsto_const_nhds, },
+  { refine measure.finite_spanning_sets_in.sigma_finite _ _,
+    { exact set.range (λ n, tᶜ ∪ T n), },
+    { refine ⟨λ n, tᶜ ∪ T n, λ n, set.mem_range_self _, λ n, _, _⟩,
+      { rw [measure.restrict_apply' (measurable_set.Union hT_meas), set.union_inter_distrib_right,
+          set.compl_inter_self t, set.empty_union],
+        exact (measure_mono (set.inter_subset_left _ _)).trans_lt (hT_lt_top n), },
+      rw ← set.union_Union tᶜ T,
+      exact set.compl_union_self _, },
+    { intros s hs,
+      rw set.mem_range at hs,
+      cases hs with n hsn,
+      rw ← hsn,
+      exact (measurable_set.compl (measurable_set.Union hT_meas)).union (hT_meas n), }, },
+end
+
+/-- A finitely strongly measurable function is measurable. -/
+protected lemma measurable [metric_space β] [measurable_space β] [borel_space β]
+  (hf : fin_strongly_measurable f μ) :
+  measurable f :=
+measurable_of_tendsto_metric (λ n, (hf.some n).measurable) (tendsto_pi.mpr hf.some_spec.2)
+
+end fin_strongly_measurable
+
+namespace ae_fin_strongly_measurable
+
+variables {α β : Type*} {m : measurable_space α} {μ : measure α} [has_zero β] [topological_space β]
+  [t2_space β] {f : α → β}
+
+lemma exists_set_sigma_finite (hf : ae_fin_strongly_measurable f μ) :
+  ∃ t, measurable_set t ∧ f =ᵐ[μ.restrict tᶜ] 0 ∧ sigma_finite (μ.restrict t) :=
+begin
+  rcases hf with ⟨g, hg, hfg⟩,
+  obtain ⟨t, ht, hgt_zero, htμ⟩ := hg.exists_set_sigma_finite,
+  refine ⟨t, ht, _, htμ⟩,
+  refine eventually_eq.trans (ae_restrict_of_ae hfg) _,
+  rw [eventually_eq, ae_restrict_iff' ht.compl],
+  exact eventually_of_forall hgt_zero,
+end
+
+/-- A measurable set `t` such that `f =ᵐ[μ.restrict tᶜ] 0` and `sigma_finite (μ.restrict t)`. -/
+def sigma_finite_set (hf : ae_fin_strongly_measurable f μ) : set α :=
+hf.exists_set_sigma_finite.some
+
+protected lemma measurable_set (hf : ae_fin_strongly_measurable f μ) :
+  measurable_set hf.sigma_finite_set :=
+hf.exists_set_sigma_finite.some_spec.1
+
+lemma ae_eq_zero_compl (hf : ae_fin_strongly_measurable f μ) :
+  f =ᵐ[μ.restrict hf.sigma_finite_setᶜ] 0 :=
+hf.exists_set_sigma_finite.some_spec.2.1
+
+lemma sigma_finite_restrict (hf : ae_fin_strongly_measurable f μ) :
```

Comment:
couldn't this even be an instance?

Reply:
I made it an instance, but I don't really understand what it brings. Since there is an explicit argument, that instance cannot be found without explicitly calling that lemma? How is it then different from a lemma?

Reply:
The conclusion of the instance is `sigma_finite (μ.restrict hf.sigma_finite_set)`. Since it mentions `hf`, this means that `hf` is available when trying to apply the instance, so everything should work fine with this instance.